### PR TITLE
test: Add test of independence of pyhf.infer from pyhf details

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

this is a test that pyhf.infer.hypotest is really independent of pyhf. This is basically preparing a out-sourcing of the inference package

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add test that pyhf.infer.hypotest is independent of pyhf to ensure that inference is decoupled
```

### CODE REVIEW

**Comment by lukasheinrich** [comment]
I guess we could have abstract base classes

```
class _configApi(object):
    def suggested_init(self) -> 'tensor':
        raise NotImplementedError('not implemented')

    def suggested_bounds(self) -> 'tensor':
        raise NotImplementedError('not implemented')

class _modelApi(object):
    def expected_data(self,pars: 'tensor', include_auxdata=True) -> 'tensor':
        raise NotImplementedError('not implemented')

    def logpdf(self, pars: 'tensor', data: 'tensor') -> 'tensor':
        raise NotImplementedError('not implemented')

```

File: tests/test_infer.py [hunk 1]
```diff
@@ -85,3 +85,61 @@ def test_hypotest_return_expected_set(tmpdir, hypotest_args):
...
+            return tensorlib.concatenate([expected_main, aux_data])
+
+        def logpdf(self, pars, data):
+            tensorlib, _ = get_backend()
+            maindata, auxdata = data
+            main = self._make_main_pdf(pars).log_prob(maindata)
+            constraint = self._make_constraint_pdf(pars).log_prob(auxdata)
+            return tensorlib.astensor([main + constraint])
+
+    m = NonPyhfModel([5, 50, 7])
```
Comment (matthewfeickert) [inline tests/test_infer.py]:
```suggestion
    model = NonPyhfModel([5, 50, 7])
```

Prefer to not have single letter variables. :)

File: tests/test_infer.py [hunk 2]
```diff
@@ -85,3 +85,61 @@ def test_hypotest_return_expected_set(tmpdir, hypotest_args):
...
+    print(pyhf.infer.hypotest(1.0, m.expected_data(m.config.suggested_init()), m))
```
Comment (matthewfeickert) [inline tests/test_infer.py]:
```suggestion
    print(pyhf.infer.hypotest(1.0, model.expected_data(model.config.suggested_init()), model))
```

Comment (matthewfeickert) [inline tests/test_infer.py]: Why use `print` here instead of an `assert`?

**Comment by kratsg** [comment]
Overall, this is mostly just using python's duck-typing here. Not that it's necessarily a bad thing and I only have a minor concern that this test is not as useful as it appears to be... that said, more tests are always good :)

**Pull request merged**